### ISSUE TITLE
feat: implement llama.cpp backend and routing classifier fine-tuning support

### ISSUE THREAD

**Issue by nikolareljin** [nikolareljin/finetorch#21]
## Overview

Implement the `llama_cpp` backend (currently a placeholder in `src/model/llama_cpp.rs`) and add a `routing` fine-tuning preset so finetorch can train small 1B–3B classification models that act as AI routing decision engines.

---

## Full Pipeline (Cross-Repo)

This feature is the **first stage** of a three-repo pipeline:

```
[finetorch]   ◄─── THIS REPO
    │               Fine-tune 1B–3B model on routing classification tasks
    │               LoRA/QLoRA → adapter.safetensors → finetorch merge → HF model dir
    ▼
[shrink-llm]        Quantize + convert to GGUF (GPTQ INT4 → Q4_K_M.gguf)
    │               routing_pipeline.yaml → models/routing/routing.Q4_K_M.gguf
    ▼
[llama-server       Serve GGUF locally on port 8080
 / bitnet-server]   OpenAI-compatible /v1/chat/completions
    ▼
[agentvault]        llm-router mode selects best AI agent per prompt
```

**Related issues:**
- `nikolareljin/agentvault` — _feat: llm-router mode_ (end consumer)
- `nikolareljin/shrink-llm` — _feat: routing task type + GGUF output pipeline_ (next step after this)

---

## Motivation

agentvault needs a locally-serving routing model to classify prompts and select the optimal AI agent. A fine-tuned 1B-3B model (LoRA/QLoRA) is ideal: small enough to run on CPU, but instruction-tuned specifically for the routing classification task.

finetorch v0.1.0 has the right architecture for this, but the critical pieces are missing:
1. `src/model/llama_cpp.rs` — `load_model`, `generate`, `apply_lora`, `save_adapter` are all stubs/`todo!()`
2. No routing/classification dataset format or preset config
3. No `merge` subcommand to produce a HF model directory for shrink-llm

---

## Current State

| Component | Status |
|-----------|--------|
| `LlmBackend` trait (`src/model/backend.rs`) | ✅ Defined |
| `LlamaCppBackend` struct (`src/model/llama_cpp.rs`) | ⚠️ Scaffold only — stubs |
| `LlamaModel::load_from_file` | ❌ Not called |
| `generate()` | ❌ Returns placeholder |
| `apply_lora()` | ❌ Not implemented |
| `save_adapter()` | ❌ Not implemented |
| Routing dataset format | ❌ Missing |
| `configs/routing_finetune.toml` | ❌ Missing |
| `finetorch merge` subcommand | ❌ Missing |

---

### 1. Implement `LlamaCppBackend` in `src/model/llama_cpp.rs`

**Add dependency** to `Cargo.toml`:

```toml
[dependencies]
llama-cpp-2 = "0.1"           # Rust bindings for llama.cpp
safetensors = "0.4"           # adapter serialization
```

**Implement the trait:**

```rust
use llama_cpp_2::{
    model::{LlamaModel, LlamaModelParams},
    context::{LlamaContext, LlamaContextParams},
    token::LlamaToken,
};
use safetensors::SafeTensors;
use std::path::{Path, PathBuf};

pub struct LlamaCppBackend {
    model:         Option<LlamaModel>,
    context:       Option<LlamaContext>,
    lora_adapters: Vec<PathBuf>,
}

impl LlamaCppBackend {
    pub fn new() -> Self {
        Self { model: None, context: None, lora_adapters: Vec::new() }
    }
}

impl LlmBackend for LlamaCppBackend {
    fn load_model(&mut self, path: &Path, config: &ModelConfig) -> Result<()> {
        if !path.exists() {
            return Err(FineError::ModelNotFound(path.to_path_buf()));
        }
        let params = LlamaModelParams::default()
            .with_n_gpu_layers(config.n_gpu_layers.unwrap_or(0));
        self.model = Some(LlamaModel::load_from_file(path, params)
            .map_err(|e| FineError::BackendError(e.to_string()))?);
        Ok(())
    }

    fn generate(&self, prompt: &str, max_tokens: usize) -> Result<String> {
        let model = self.model.as_ref()
            .ok_or(FineError::ModelNotLoaded)?;
        let ctx_params = LlamaContextParams::default()
            .with_n_ctx(NonZeroU32::new(2048).unwrap());
        let mut ctx = model.new_context(ctx_params)
            .map_err(|e| FineError::BackendError(e.to_string()))?;

        // Tokenize, run decode loop, collect output tokens
        let tokens = model.str_to_token(prompt, AddBos::Always)
            .map_err(|e| FineError::BackendError(e.to_string()))?;
        // ... inference loop up to max_tokens
        // Return decoded string
        todo!("inference loop — implement token sampling and detokenization")
    }

    fn apply_lora(&mut self, adapter_path: &Path) -> Result<()> {
        let model = self.model.as_ref().ok_or(FineError::ModelNotLoaded)?;
        model.apply_lora_from_file(adapter_path, 1.0, None)
            .map_err(|e| FineError::BackendError(e.to_string()))?;
        self.lora_adapters.push(adapter_path.to_path_buf());
        Ok(())
    }

    fn save_adapter(&self, output_path: &Path) -> Result<()> {
        // Serialize LoRA adapter weight deltas as safetensors
        // Requires extracting delta weights from model — see safetensors crate
        todo!("extract and serialize LoRA weight deltas")
    }
}
```

**Note on `generate` and `save_adapter`**: These require non-trivial inference loop and weight extraction logic. Implement incrementally — `load_model` and `apply_lora` are the critical path for fine-tuning; `generate` is needed for eval; `save_adapter` for export.

### 2. Routing dataset format

Target: `datasets/routing/train.jsonl` (200 examples) and `datasets/routing/eval.jsonl` (50 examples)

```jsonl
{"instruction": "Route this task: 'write a binary search tree in Rust'", "response": "{\"task_type\": \"coding\", \"complexity\": 6, \"requires_tools\": false, \"privacy_sensitive\": false}"}
{"instruction": "Route this task: 'review this authentication module for vulnerabilities'", "response": "{\"task_type\": \"review\", \"complexity\": 7, \"requires_tools\": false, \"privacy_sensitive\": true}"}
{"instruction": "Route this task: 'explain the CAP theorem'", "response": "{\"task_type\": \"analysis\", \"complexity\": 4, \"requires_tools\": false, \"privacy_sensitive\": false}"}
{"instruction": "Route this task: 'what time is it'", "response": "{\"task_type\": \"general\", \"complexity\": 1, \"requires_tools\": false, \"privacy_sensitive\": false}"}
```

The `instruction` format must match the system prompt used by agentvault's llm-router (routing factors JSON).

### 3. New config: `configs/routing_finetune.toml`

```toml
[model]
base    = "meta-llama/Llama-3.2-1B-Instruct"
backend = "llama_cpp"

[lora]
enabled        = true
use_qlora      = true
r              = 8
alpha          = 16
dropout        = 0.05
target_modules = ["q_proj", "v_proj"]

[training]
epochs                     = 3
batch_size                 = 4
learning_rate              = 2e-4
max_seq_len                = 512
gradient_accumulation_steps = 4
warmup_steps               = 50

[data]
train  = "datasets/routing/train.jsonl"
eval   = "datasets/routing/eval.jsonl"
format = "instruction"   # {"instruction": "...", "response": "..."}

[eval]
metrics        = ["exact_match", "bleu"]
eval_every_n_steps = 100

[output]
adapter_path = "artifacts/routing-lora/adapter.safetensors"
```

### 4. New subcommand: `finetorch merge`

New file: `src/cli/merge.rs`

```rust
/// Merges a LoRA adapter into a base model and writes a HuggingFace model directory.
/// Output is passed to shrink-llm's gguf_convert stage.
pub fn run_merge(args: &MergeArgs) -> Result<()> {
    // 1. Load base model
    // 2. Load adapter (safetensors)
    // 3. Add adapter weights to base model weight tensors
    // 4. Write merged weights + config.json + tokenizer.json to output_dir
    // Output is a standard HF model directory loadable by shrink-llm
}

#[derive(Args)]
pub struct MergeArgs {
    #[arg(long)] base:    String,   // HF model ID or local path
    #[arg(long)] adapter: PathBuf, // path to adapter.safetensors
    #[arg(long)] output:  PathBuf, // output directory
}
```

CLI usage:
```bash
finetorch merge \
  --base meta-llama/Llama-3.2-1B-Instruct \
  --adapter artifacts/routing-lora/adapter.safetensors \
  --output artifacts/routing-merged/
```

Add `merge` to `src/main.rs` subcommand dispatch.

### 5. Tests

`tests/test_llama_cpp_backend.rs`:

```rust
#[test]
fn test_load_model_missing_file_returns_error() {
    let mut backend = LlamaCppBackend::new();
    let result = backend.load_model(
        Path::new("/nonexistent/model.gguf"),
        &ModelConfig::default(),
    );
    assert!(result.is_err());
    let err = result.unwrap_err().to_string();
    assert!(err.contains("not found") || err.contains("No such file"),
            "unexpected error: {err}");
}

#[test]
fn test_generate_without_loaded_model_returns_error() {
    let backend = LlamaCppBackend::new();
    let result = backend.generate("hello", 10);
    assert!(result.is_err());
}

#[test]
fn test_apply_lora_without_model_returns_error() {
    let mut backend = LlamaCppBackend::new();
    let result = backend.apply_lora(Path::new("/fake/adapter.safetensors"));
    assert!(result.is_err());
}
```

`tests/test_routing_config.rs`:

```rust
#[test]
fn test_routing_config_parses_correctly() {
    let cfg = FinetorchConfig::from_file("configs/routing_finetune.toml")
        .expect("config should parse");
    assert_eq!(cfg.model.backend, "llama_cpp");
    assert!(cfg.lora.enabled);
    assert_eq!(cfg.lora.r, 8);
    assert_eq!(cfg.training.epochs, 3);
    assert_eq!(cfg.data.format, "instruction");
}

#[test]
fn test_merge_args_parse() {
    // Verify clap can parse merge subcommand args
    let app = build_cli();
    let m = app.try_get_matches_from([
        "finetorch", "merge",
        "--base", "meta-llama/Llama-3.2-1B-Instruct",
        "--adapter", "adapter.safetensors",
        "--output", "/tmp/out",
    ]);
    assert!(m.is_ok());
}
```

---

## Implementation Order

1. Add `llama-cpp-2` and `safetensors` to `Cargo.toml`
2. Implement `LlamaCppBackend::load_model` — file existence check + `LlamaModel::load_from_file`
3. Implement `LlamaCppBackend::apply_lora` — `model.apply_lora_from_file`
4. Write tests: `test_load_model_missing_file_returns_error`, `test_apply_lora_without_model_returns_error`
5. Implement `generate()` inference loop (tokenize → decode → detokenize)
6. Implement `save_adapter()` weight delta extraction + safetensors serialization
7. Create `datasets/routing/train.jsonl` (200 examples) and `eval.jsonl` (50 examples)
8. Create `configs/routing_finetune.toml`
9. Create `src/cli/merge.rs` + `MergeArgs` struct
10. Add `merge` to subcommand dispatch in `src/main.rs`
11. Write tests: `test_routing_config_parses_correctly`, `test_merge_args_parse`
12. Update `docs/backends.md` with llama_cpp usage instructions

---

## Acceptance Criteria

- [ ] `finetorch train --config configs/routing_finetune.toml` runs without panic on CPU (even if slow — real gradient training not required for v1, forward pass suffices)
- [ ] `artifacts/routing-lora/adapter.safetensors` is created after training completes
- [ ] `finetorch merge --base ... --adapter ... --output ...` writes a directory with `config.json` + weight files
- [ ] `cargo test` — all existing tests still pass
- [ ] New tests pass: `cargo test test_llama_cpp_backend test_routing_config`
- [ ] `cargo clippy -- -D warnings` — clean

---

## Notes

- **Real gradient training**: The existing training loop in `src/train/` uses synthetic batches. Wiring real dataset loading + actual gradient steps (via `llama-cpp-2` or Candle) is the hard part and may be phased. Priority: get `load_model`, `apply_lora`, `generate` working for forward-pass eval first.
- **GGUF input**: The `llama_cpp` backend loads `.gguf` files directly. Users can download pre-quantized GGUF models (e.g., `Llama-3.2-1B-Instruct-Q4_K_M.gguf`) for fast iteration without running through shrink-llm.
- **Merge output format**: The `merge` command output must be a standard HuggingFace model directory (with `config.json`, `tokenizer.json`, and weight files) so shrink-llm's `gguf_convert` stage can process it with `convert-hf-to-gguf.py`.
